### PR TITLE
Backend paddle: add regularizer

### PR DESCRIPTION
**Pull request by lijialin03**
1. add l1/l2 regularizer for paddle backend
2. refactor regularizer to be compatible with multiple backends

### CODE REVIEW

File: deepxde/backend/backend.py [hunk 1]
```diff
@@ -502,3 +502,12 @@ def sparse_dense_matmul(x, y):
     Returns:
         Tensor: The multiplication result.
     """
+
```
Comment (lululxvi) [inline deepxde/backend/backend.py:505]:
```
###############################################################################
# Regularization
```

File: deepxde/backend/backend.py [hunk 2]
```diff
@@ -502,3 +502,19 @@ def sparse_dense_matmul(x, y):
...
+
+###############################################################################
+# Regularization
+
+
+def l1_decay(x):
```
Comment (lululxvi) [inline deepxde/backend/backend.py:511]:
```
l1_regularization
```

File: deepxde/backend/backend.py [hunk 3]
```diff
@@ -502,3 +502,19 @@ def sparse_dense_matmul(x, y):
...
+    """Implement the L1 weight decay regularization."""
+
+
+def l2_decay(x):
```
Comment (lululxvi) [inline deepxde/backend/backend.py:515]:
```
l2_regularization
```

File: deepxde/backend/backend.py [hunk 4]
```diff
@@ -502,3 +502,19 @@ def sparse_dense_matmul(x, y):
...
+    """Implement the L2 weight decay regularization."""
+
+
+def l1_l2_decay(x,y):
```
Comment (lululxvi) [inline deepxde/backend/backend.py:519]:
```
l1_l2_regularization
```

File: deepxde/backend/backend.py (hunk 2 above)
Comment (lululxvi) [inline deepxde/backend/backend.py:511]: `l1_regularization`

File: deepxde/backend/backend.py (hunk 3 above)
Comment (lululxvi) [inline deepxde/backend/backend.py:515]: `l2_regularization`

File: deepxde/backend/backend.py (hunk 4 above)
Comment (lululxvi) [inline deepxde/backend/backend.py:519]:
```
l1_l2_regularization
```

File: deepxde/backend/backend.py [hunk 5]
```diff
@@ -502,3 +502,19 @@ def sparse_dense_matmul(x, y):
...
         Tensor: The multiplication result.
     """
+
+
+###############################################################################
+# Regularization
+
+
+def l1_regularization(x):
+    """Implement the L1 weight decay regularization."""
```
Comment (lululxvi) [inline deepxde/backend/backend.py:512]:
```
def l1_regularization(l1):
    """A regularizer that applies a L1 regularization penalty or L1 weight decay."""
```

File: deepxde/backend/backend.py [hunk 6]
```diff
@@ -502,3 +502,19 @@ def sparse_dense_matmul(x, y):
...
+
+
+def l2_regularization(x):
+    """Implement the L2 weight decay regularization."""
```
Comment (lululxvi) [inline deepxde/backend/backend.py:516]:
```
def l2_regularization(l2):
    """A regularizer that applies a L2 regularization penalty or L2 weight decay."""
```

File: deepxde/backend/backend.py [hunk 7]
```diff
@@ -502,3 +502,19 @@ def sparse_dense_matmul(x, y):
...
+
+
+def l1_l2_regularization(x,y):
+    """Implement the L1 and L2 weight decay regularization."""
```
Comment (lululxvi) [inline deepxde/backend/backend.py:520]:
```
def l1_l2_regularization(l1, l2):
    """A regularizer that applies both L1 and L2 regularization penalties or L1 and L2 weight decay."""
```

File: deepxde/backend/backend.py [hunk 8]
```diff
@@ -502,3 +502,41 @@ def sparse_dense_matmul(x, y):
...
+    Warning:
+        The implementation may vary across different backends.
+
+    Args:
+        l2 (float): L2 regularization factor.
+    """
+
+
+def l1_l2_regularization(l1, l2):
+    """A regularizer that applies both L1 and L2 regularization penalties or L1 and L2 weight decay.
```
Comment (lululxvi) [inline deepxde/backend/backend.py:534]: max 88 characters per line

File: deepxde/backend/tensorflow_compat_v1/tensor.py [hunk 9]
```diff
@@ -245,3 +245,15 @@ def matmul(x, y):
...
+
+def l1_regularization(x):
+    return tf.keras.regularizers.L1(l1=x)
+
+
+def l2_regularization(x):
+    return tf.keras.regularizers.L2(l2=x)
+
+
+def l1_l2_regularization(x,y):
```
Comment (lululxvi) [inline deepxde/backend/tensorflow_compat_v1/tensor.py:258]:
```
def l1_l2_regularization(x, y):
```

File: deepxde/nn/regularizers.py [hunk 10]
```diff
@@ -22,12 +23,15 @@ def get(identifier):
...
 
-    if name == "l1":
-        return tf.keras.regularizers.L1(l1=factor[0])
-    if name == "l2":
-        return tf.keras.regularizers.L2(l2=factor[0])
-    if name in ("l1l2", "l1+l2"):
-        if len(factor) < 2:
-            raise ValueError("L1L2 regularizer requires both L1/L2 penalties.")
-        return tf.keras.regularizers.L1L2(l1=factor[0], l2=factor[1])
+    try:
```
Comment (lululxvi) [inline deepxde/nn/regularizers.py:26]: `try` is not needed, see
https://github.com/lululxvi/deepxde/blob/8275aeb6792709f51c9ebc3a7655cc0f20623c1c/deepxde/backend/__init__.py#L27

File: deepxde/backend/paddle/tensor.py [hunk 11]
```diff
@@ -229,3 +229,11 @@ def matmul(x, y):
 
 def sparse_dense_matmul(x, y):
     return paddle.sparse.matmul(x, y)
+
+
+def l1_regularization(x):
```
Comment (lululxvi) [inline deepxde/backend/paddle/tensor.py:234]: use l1 or l2 to replace x.

File: deepxde/nn/regularizers.py [hunk 12]
```diff
@@ -23,11 +23,11 @@ def get(identifier):
...
+        return bkd.l1_regularization(factor[0])
     if name == "l2":
-        return tf.keras.regularizers.L2(l2=factor[0])
+        return bkd.l2_regularization(factor[0])
     if name in ("l1l2", "l1+l2"):
+        # TODO: only supported by 'tensorflow.compat.v1' now.
         if len(factor) < 2:
-            raise ValueError("L1L2 regularizer requires both L1/L2 penalties.")
-        return tf.keras.regularizers.L1L2(l1=factor[0], l2=factor[1])
+            return bkd.l1_l2_regularization(factor[0], factor[1])
```
Comment (lululxvi) [inline deepxde/nn/regularizers.py:32]: A mistake here.

**Comment by lijialin03** [comment]
> Also need to add the regularization functions to the tensorflow backend

Ok, I want to confirm, are you referring to 'tensorflow' and not 'tensorflow.compat.v1'? I mean regularization functions have been added to 'tensorflow.compat.v1' now.

File: deepxde/nn/regularizers.py [hunk 13]
```diff
@@ -23,11 +23,12 @@ def get(identifier):
         raise ValueError("Regularization factor must be provided.")
 
     if name == "l1":
-        return tf.keras.regularizers.L1(l1=factor[0])
+        return bkd.l1_regularization(factor[0])
     if name == "l2":
-        return tf.keras.regularizers.L2(l2=factor[0])
+        return bkd.l2_regularization(factor[0])
     if name in ("l1l2", "l1+l2"):
+        # TODO: only supported by 'tensorflow.compat.v1' now.
```
Comment (lululxvi) [inline deepxde/nn/regularizers.py:30]: Need to update the docstring in line 5.

File: deepxde/nn/regularizers.py [hunk 14]
```diff
@@ -23,11 +23,12 @@ def get(identifier):
         raise ValueError("Regularization factor must be provided.")
 
     if name == "l1":
-        return tf.keras.regularizers.L1(l1=factor[0])
+        return bkd.l1_regularization(factor[0])
     if name == "l2":
-        return tf.keras.regularizers.L2(l2=factor[0])
+        return bkd.l2_regularization(factor[0])
     if name in ("l1l2", "l1+l2"):
+        # TODO: only supported by 'tensorflow.compat.v1' and 'tensorflow' now.
```
Comment (lululxvi) [inline deepxde/nn/regularizers.py:30]: Delete line 14.

**Pull request merged**